Model Pipeline + Generation Pipeline

vineetbansal opened the pull request:
1. Model Pipeline + Generation Pipeline
2. Regularization support for mean/covariance estimation
3. Simplification of metadata field handling by using Relion labels everywhere
4. Support for default values in get_metadata method of ImageSource for missing metadata fields.

Right now this is a single squashed commit, just so we can get things going. I can separate this out into 4 commits if this is getting too mixed up.

vineetbansal committed:
Model Pipeline + Generation Pipeline

Also regularization support for mean/covariance estimation; Simplification of metadata field handling by using Relion labels everywhere; support for default values in get_metadata method of ImageSource for missing metadata fields.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -171,19 +235,43 @@ def set_metadata(self, metadata_fields, values, indices=None):
+                else:
+                    result = right
+            else:
+                raise RuntimeError('Missing columns and no default value provided!')
```
Let's avoid exclamation points in messages.

vineetbansal replied:
Removed

janden commented on src/aspire/estimation/__init__.py:
```diff
@@ -49,10 +52,10 @@ def __getattr__(self, name):
         raise NotImplementedError('Subclasses must implement the compute_kernel method')
 
-    def estimate(self, b_coeff=None):
+    def estimate(self, b_coeff=None, tol=None, regularizer=None):
```
How does the regularizer work? Is it just a scalar value? Might be good to document this.

vineetbansal replied:
Documentaion will be added en masse soon.

janden commented on src/aspire/estimation/__init__.py:
```diff
@@ -67,25 +70,39 @@ def src_backward(self):
+
+        if regularizer is None:
+            regularizer = config.mean.regularizer
+        if regularizer > 0:
```
Is this `if` statement needed?

vineetbansal replied:
The extra `if` was to have the caller either explicitly specify `regularizer`, or rely on the `config` to pick up values (which in turn could be `0` or a +ve value). Probably enough just to be able to specify in `config` (which can be overriden in scripts anyway). Removed now.

janden replied:
I guess I had a more specific question: if `regularizer` is zero, adding it to the kernel is a no-op, so why the `if` statement?

janden commented on src/aspire/estimation/__init__.py:
```diff
@@ -67,25 +70,39 @@ def src_backward(self):
         else:
-            M = LinearOperator((n, n), matvec=partial(self.apply_kernel, kernel=self.precond_kernel))
+            precond_kernel = self.precond_kernel
+            if regularizer > 0:
```
See above. I don't think the `if` clause is needed here.

vineetbansal replied:
The extra if was to have the caller either explicitly specify regularizer, or rely on the config to pick up values (which in turn could be 0 or a +ve value). Probably enough just to be able to specify in config (which can be overriden in scripts anyway). Removed now

janden commented on src/aspire/estimation/covar.py:
```diff
@@ -85,10 +86,10 @@ def compute_kernel(self):
         return FourierKernel(kernel_f, centered=False)
 
-    def estimate(self, mean_vol, noise_variance, tol=None):
+    def estimate(self, mean_vol, noise_variance, tol=None, regularizer=None):
```
Again, might want to briefly document this since it's getting more complicated.

vineetbansal replied:
The `regularizer` argument no longer exists. Documenting functions is something I intend to do in the next couple months when I'm no longer officially on the project. Closing here, as having this issue may give the false impression that this is the only place where documention is missing.

janden replied:
Ok good.

janden commented on src/aspire/estimation/covar.py:
```diff
@@ -97,17 +98,24 @@ def estimate(self, mean_vol, noise_variance, tol=None):
+
+        if regularizer is None:
+            regularizer = config.covar.regularizer
+        if regularizer > 0:
```
`if` clause needed?

vineetbansal replied:
Gone

janden commented on src/aspire/estimation/covar.py:
```diff
@@ -97,17 +98,24 @@ def estimate(self, mean_vol, noise_variance, tol=None):
         else:
-            M = LinearOperator((N, N), matvec=partial(self.apply_kernel, kernel=self.precond_kernel, packed=True))
+            precond_kernel = self.precond_kernel
+            if regularizer > 0:
```
`if` clause needed?

vineetbansal replied:
gone

janden commented on src/aspire/estimation/kernel.py:
```diff
@@ -24,6 +24,19 @@ def __init__(self, kernel, centered):
 
+    def __add__(self, regularizer):
+        """
+        There is often a need to add a regularization parameter (a small positive value) to a FourierKernel object,
```
Might want to reformulate this a bit. Adding a constant to a kernel could be due to regularization, but it's a rather general operation, so let's leave it at that level.

vineetbansal replied:
Not sure I understand..did you mean to say just make it a generic `add` operation without mentioning regularization (and changing argument names), or not have the operation at all (i.e. allowing the underlying `ndarray` to be modified). The latter scenario won't quite work without refactoring a bunch of things (since the `FourierKernel` object is remembered beyond this optimization procedure and needs to remember its `kernel` value without modifications).

janden replied:
> did you mean to say just make it a generic add operation without mentioning regularization (and changing argument names)

Yes.

vineetbansal replied:
Makes sense..modified the method signature/docstrings to use `delta` instead of `regularizer`. Added a note about why this method even exists, in a `note` docstring directive.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -170,6 +174,10 @@ def filter(self, filter):
         raise NotImplementedError
 
+    def multiply(self, amplitudes):
+        amplitudes = np.broadcast_to(amplitudes, (self.res, self.res, len(amplitudes)))
```
This generates a large intermediate array, correct? Why not perform the broadcasting as part of the multiplication? Should be more efficient.

vineetbansal replied:
Yes ... not sure why I did this. Must be the hashish I was smoking. Fixed now.

janden commented on src/aspire/utils/filters.py:
```diff
@@ -9,6 +10,19 @@
 class Filter:
+
+    @classmethod
+    def from_func(cls, f, dim=None):
```
Why not make this a new subclass (say `FunctionFilter`)? Seems a little sneaky to instantiate the base class and modify the functions in-place. Another option is to make this the basic functionality of the `Filter` class and have subclasses override it as needed.

vineetbansal replied:
New class `FunctionFilter` added.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -149,13 +152,14 @@ def downsample(self, ds_res):
         :return: A new filtered Image object.
         """
-        filter_values = filter.evaluate_grid(self.res)
+        filter_values = filter.evaluate_grid(self.res) ** power
```
This creates two ways of filtering by a certain power of a filter: modifying its `power` attribute or sending a different parameter to the `Image.filter` method. This seems redundant.

vineetbansal replied:
Well, I thought about this for a while, and there are 2 complementary scenarios, both of which are used in different parts of the code.

1) Filters have to be evaluated at a particular power on a one-off basis, without modifying them (which would create havoc). For example, the mean estimator and covariance estimator ask the `ImageSource` to give the results of the evaluation of its filters at `power=2`.

Another place is when a `NoiseAdder` applies a `noise_filter` to generated noise (`power=0.5`).

2) The power of the filter needs to be stored inside the filter itself. E.g. the `noise_psd` returned by a `NoiseEstimator` is stored as a filter with power -0.5 in the generation pipeline.

There's probably a way to collapse both use cases into one, but I'm not seeing it currently. If (2) above is only ever used with `ArrayFilter` objects, then we could probably just construct a new one with squared values of xfer_function and get rid of the `power` attribute.

vineetbansal replied:
Need to add a new `PowerFilter` class and utilize it in these 2 cases. Remove `power` attribute and argument from appropriate places.

vineetbansal replied:
`PowerFilter` added. `power` removed as an attribute of a `Filter` object. `power` argument removed from the `Image.filter` method.

janden commented on src/aspire/utils/filters.py:
```diff
@@ -181,19 +198,18 @@ def __init__(self, pixel_size=None, voltage=None, defocus_u=None, defocus_v=None
         :param alpha:       Amplitude contrast phase in radians
-        :param B:           Envelope decay in inverse square Angstrom
-        :param power:       Power of the filter (default 1)
+        :param B:           Envelope decay in inverse square Angstrom (default 0)
```
Lowercase “angstrom” (also for the above).

vineetbansal replied:
"Angstrom" lowercased. `Power` issue discussed in another conversation, so closing here.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        super().downsample(resolution)
+
+
+class DownSample(OneWayXform):
```
“Downsample” is one word, so in CamelCase, that's `Downsample`.

Also, note that downsampling is a linear operation, so it does have an adjoint (and it is quite useful in certain situations). The adjoint consists of inserting zeros at all points that aren't in the downsampling grid. In other words, if downsampling takes us from `[2, 3, 5, 7, 11, 13, 17, 19]` to `[2, 5, 11, 17]`, its adjoint (known as upsampling) takes `[23, 27, 29, 31]` and gives `[23, 0, 27, 0, 29, 0, 31]`. It might be useful to have this `upsample` method in the `Image` class and have `Downsample._adjoint` call it. Not necessary for this PR, but might be something to think about.

vineetbansal replied:
Upsampling support added as a new issue #71 , so closing here. Changed DownSample -> Downsample.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+logger = logging.getLogger(__name__)
+
+
+class Xform:
```
I think the class structure is a bit upside-down here. The presence/absence of an adjoint has to do with the linearity of the transform. In other words, a general mapping has a forward, but not necessarily an adjoint. A linear mapping is a subset of a general mapping, but since it is linear, it is guaranteed to have an adjoint. A more sensible hierarchy might then be `Xform` → `LinearXform`, where `Xform` has `forward` and `LinearXform` has both `forward` and `adjoint`. Furthermore, we'd have `LinearXform` → `SymmetricXform`. The other subclasses would then inherit from `LinearXform` or `SymmetricXform`, except for `NoiseAdder`, which is non-linear and would inherit from `Xform`.

vineetbansal replied:
Excellent! There are some code-glueing issues that call for multiple-inheritence (a `Pipeline` just extends a `Xform`, while a `LinearPipeline` would extend both `Pipeline` and `LinearXform`), but the code itself after this refactoring is straightforward to understand.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+    Xform objects usually set up data structures that are typically *larger* than the depth of the Image
+    object that they expect to encounter during any invocation of `forward` or `adjoint`.
+
+    At runtime, it gets an Image object (a thin wrapper on a L x L x n ndarray),
```
“x” → “×”.

vineetbansal replied:
The `x` thing for multiplication appears at a lot of places. I'm a little wary of introducing non-ascii characters in code. Let me open up a new issue on this to fix en masse or ignore altogether. Closing here.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+    At runtime, it gets an Image object (a thin wrapper on a L x L x n ndarray),
+    as well as numeric `indices` (a numpy array of index values) that correspond to index values (the 'window') of
+    the incoming Image object within the context of all images (i.e. an Image object of L x L x N, where N represents
```
“x” → “×”. Also, let's keep the line width under 80 characters for the docstrings.

vineetbansal replied:
“x” → “×” opened as new issue #70 . The pep8 character width issue opened as #69. Closing here.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+    At runtime, it gets an Image object (a thin wrapper on a L x L x n ndarray),
+    as well as numeric `indices` (a numpy array of index values) that correspond to index values (the 'window') of
+    the incoming Image object within the context of all images (i.e. an Image object of L x L x N, where N represents
+    the no. of total images that can ever pass through this Xform.
```
“no.” → “number”.

vineetbansal replied:
fixed. I think `no.` is an Indianism that I need to get over.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+    class XformActiveContextManager:
+        """
+        This inner class allows us to temporarily enable/disable a Xform object, by tweaking it's `active`
```
“it's” → “its”.

vineetbansal replied:
Fixed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+    def __init__(self, resolution=np.inf, active=True):
+        """
+        Create a Xform object that works at a specific resolution.
+        :param resolution: The side-length of the Image object on which this Xform is applicable.
```
Could we have a quick note of what the special value `np.inf` means here?

vineetbansal replied:
Change to `None` to avoid confusion.

vineetbansal replied:
`resolution` not maintained as an attribute anymore, thus removed as constructor argument.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+            indices = np.arange(im.n_images)
+        return self._adjoint(im, indices=indices)
+
+    def _adjoint(self, im, indices):
```
Again, might want to forgo this for the general (possibly non-linear) `Xform` and only require it in `LinearXform`.

vineetbansal replied:
Method removed from `Xform` (applicable only to `LinearXform`)

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+    def downsample(self, resolution):
+        """
+        Downsample this Xform object to deal with Image objects of size resolution x resolution x n.
```
“x” → “×”.

vineetbansal replied:
Opened as #70. Closing here.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        Downsample this Xform object to deal with Image objects of size resolution x resolution x n.
+        On return, the Xform has been modified to expect Image objects in new resolution.
+        Subclasses will typically want to do more than simply make a note of the new resolution as is done here.
+        :param resolution: The new resolution of incoming Image objects in `forward` or `adjoint`
```
End with period.

vineetbansal replied:
Fixed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        return self._forward(im, indices)
+
+
+class OneWayXform(Xform):
```
Under the proposed hierarchy, this class wouldn't exist.

vineetbansal replied:
Yes. This class is now gone.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+class Multiply(SymmetricXform):
+    """
+    A Xform that changes the amplitudes of a 3D Image by multiplying all pixels of a single 2d image by a constant
```
I don't like the terminology “3D Image” since it can be confused with a 3D map. What about “a stack of 2D images in the form of an `Image` object”. Also “2d” → “2D”.

vineetbansal replied:
fixed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        """
+        Initialize a Multiply Xform using specified factors
+        :param factor: An ndarray of scalar factors to use for amplitude multiplication.
+        :param resolution: Resolution of images expected to pass through this Transform
```
“Transform” → “`Xform`”.

vineetbansal replied:
fixed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        self.multipliers = factor
+
+    def _forward(self, im, indices):
+        return Image(im.asnumpy() * self.multipliers[indices])
```
Is this not implemented by `Image.__multiply__`?

vineetbansal replied:
It's not currently, but should be. Added a `__mul__` method to the `Image` class.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+class Shift(Xform):
+    """
+    A Xform that shifts pixels of a 3D Image by offsetting all pixels of a single 2d image by constant x/y offsets.
```
See comment on `Multiply` docstring.

vineetbansal replied:
fixed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+    """
+    def __init__(self, shifts, resolution=np.inf):
+        """
+        Initialize a Shift Transform using a numpy array of shift values.
```
“Shift Transform” → “`Shift` `Xform`”. “numpy” → “NumPy”.

vineetbansal replied:
fixed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+class Filter(SymmetricXform):
+    """
+    A Xform that applies a single CTF Filter object to a 3D Image.
```
Doesn't have to be a CTF, right? Maybe just “`Filter`” would do. Also, let's say either “`Image`” or “2D image stack in the form of an `Image` object”.

vineetbansal replied:
verbiage changed.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+    """
+    def __init__(self, filter, resolution=np.inf):
+        """
+        Initialize the Filter Xform using a Filter object
```
Backticks around “Filter” and “Xform”.

vineetbansal replied:
added

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        return im.downsample(self.resolution)
+
+
+class Filter(SymmetricXform):
```
Doesn't this clash with the other `Filter` class? Maybe these should all have the suffix `Xform` to distinguish them from other objects.

Also, filtering is only symmetric if the transfer function itself is symmetric. This is the case for many of them (such as the CTF), but is not generally the case. For now, we can pretend the adjoint doesn't exist (like in `Downsample`, above), but at some point, we should have something like `Filter.dual()` that will return the dual filter (the same filter, but with a transfer function flipped around the origin, that is g(ω) = f(–ω), where f(ω) is the transfer function of the original filter).

vineetbansal replied:
changed name to `FilterXform`. Good point about the `dual` method for a `Filter`. I've added a new issue #72 for that. Closing here.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+class IndexedXform(Xform):
+    """
+    An IndexedXform is a Xform where individual Xform objects are used at specific indices of the incoming Image object.
```
Make sure docstrings have line widths below 80 characters.

vineetbansal replied:
This might be violated at a bunch of places. Let's keep this under consideration in issue #69. In general I find 80 chars too restrictive. Closing here.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+            assert np.min(indices) >= 0
+            assert np.max(indices) < len(unique_xforms)
+
+        resolution = min(np.inf, *[xform.resolution for xform in unique_xforms])
```
What behavior does this produce? Should we ever be combining `Xform` of different resolutions? It's not clear to me how to apply those to a single stack of images (which, by definition, has a single resolution).

vineetbansal replied:
This is no longer an issue since Xform do not maintain a track of their `resolution` anymore (which was only ever used to downsample them). If the `forward` and `adjoint` methods of an `Xform` can deal with an `Image` of a particular resolution at runtime, that's good enough.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        im_data = im.asnumpy().copy()
+
+        # For each individual transformation
+        for i in range(self.n_xforms):
```
Maybe `for (i, xform) in enumerate(self.unique_xforms)`?

vineetbansal replied:
Yes! changed.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        # For each individual transformation
+        for i in range(self.n_xforms):
+            # Get the indices corresponding to that transformation
+            idx = np.where(self.indices == i)[0]
```
Just out of curiosity, what does `[0]` do here?

vineetbansal replied:
`np.where` has a slightly weird behavior for a single argument (i.e. when you just want the indices of `True` values based on a condition). In this case it returns results in the 0th element of the result (which is just a 1 element array).

Reading up their documentation, it seems like in our use case (as in most cases where `np.where` is actually used), it's probably cleaner to use `np.flatnonzero`. Changed.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        # Ensure that we will be able to apply all transformers to the image
+        assert self.n_indices >= im.n_images, f'Can process Image object of max depth {self.n_indices}. Got {im.n_images}.'
+
+        im_data = im.asnumpy().copy()
```
`copy()` seems to be overkill here. What about `np.empty_like(im.asnumpy())`?

vineetbansal replied:
Good catch. Changed.

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        super().downsample(resolution)
+
+
+def _apply_transform(xform, im, indices, adjoint=False):
```
I think it would be good to just say `xform` everywhere instead of having two ways to write it.

vineetbansal replied:
changed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        else:
+            self.resolution = np.inf
+
+    def add_transform(self, xform):
```
`transform` → `xform` (see above).

vineetbansal replied:
changed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        self.xforms.append(xform)
+        self._register_xforms_changed()
+
+    def add_transforms(self, xforms):
```
`transform` → `xform` (see above).

vineetbansal replied:
changed

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+    def add_transform(self, xform):
+        """
+        Add a single Xform object at the end of the pipeline.
```
Backticks around `Xform`.

vineetbansal replied:
added

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+    def __init__(self, xforms=None, memory=None):
+        """
+        Initialize a Pipeline with Xform objects.
```
Backticks around `Pipeline` and `Xform`.

vineetbansal replied:
added

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+class Pipeline(Xform):
+    """
+    A Pipeline is a Xform made up of individual transformation steps (i.e. multiple Xform objects).
```
Make sure docstrings are shorter than 80 characters in line width. Also, backticks around class names.

vineetbansal replied:
Added backticks. Issue #69 for line width (which we can tackle in one fell swoop)

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+    def _register_xforms_changed(self):
+        # Update resolution of the Pipeline in case any of the transformations changed.
+        if len(self.xforms) > 0:
+            self.resolution = min(np.inf, *[xform.resolution for xform in self.xforms])
```
Shouldn't `Pipeline.resolution` be the resolution of the first `Xform` in the pipeline?

vineetbansal replied:
So the whole reason that currently we can add downsampling as a lazy operation (done by appending a `DownSample` Xform towards the end of the generation pipeline, is because the individual `Xform` objects in a `Pipeline` don't have to be of the same resolution (i.e. they can spit out `Image` objects in different resolution).

I'll have to think about this one.

vineetbansal replied:
We've now done away with a `Pipeline` maintaining it's `resolution` . Because we're not maintaining a `model_pipeline` anymore, there is no need to downsample each `Xform` in the `Pipeline`, and thus no need to maintain a `resolution` attribute for a `Xform` object (except for a `DownSample` Xform of course).

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+
+    def add_transforms(self, xforms):
+        """
+        Add multiple Xform objects at the end of the pipeline.
```
Backticks.

vineetbansal replied:
added

janden commented on src/aspire/source/xform.py:
```diff
@@ -0,0 +1,388 @@
+        """
+        return Xform.XformActiveContextManager(self, active=False)
+
+    def downsample(self, resolution):
```
I keep going back and forth about whether this should be in here or not. It's a bit confusing that there's a `downsample` method, but also `Downsample` class. How are they related? This may be indicative of some deeper issue.

vineetbansal replied:
With there being no model-pipeline anymore, the operation of downsampling a whole Pipeline is now gone. (We just deal with a `Downsample` Xform class).

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
-        '_alpha':       '_rlnAmplitudeContrast'
+    """
+    When creating an ImageSource object, a 'metadata' table holds metadata information about all images in the
+    ImageSource. The number of rows in this metadata table will equal the total no. of images supported by this
```
“no.” → “number”

vineetbansal replied:
changed

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
+    begin with a single underscore '_'.
+
+    In addition, the metadata table may also contain references to Python objects.
+    'Filter' objects, for example, are stored in this metadata table as references  to unique Filter objects that
```
Remove double space.

vineetbansal replied:
removed

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
-        '_Cs':          '_rlnSphericalAberration',
-        '_alpha':       '_rlnAmplitudeContrast'
+    """
+    When creating an ImageSource object, a 'metadata' table holds metadata information about all images in the
```
Backticks throughout around  class names.

vineetbansal replied:
added

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
+    The STAR file used to initialize subclasses of ImageSource may well contain other columns not found below; these
+    additional columns are available when read, and they default to the pandas data type 'object'.
+    """
+    metadata_fields = {
```
What's the motivation for using the RELION column names in our code? I thought there was a translation layer at some point where these were changed into simpler names.

vineetbansal replied:
The translation layer was causing a lot of noise (lots of forward and reverse dictionaries applied at different places) for very little gains. It's largely irrelevant since access to key properties (offsets, amplitudes) is through properties which look up the metadata field.

I propose we add back a translation layer if we ever decide to support multiple serialization formats to/from starfile (in addition to Relion).

janden replied:
> I propose we add back a translation layer if we ever decide to support multiple serialization formats to/from starfile (in addition to Relion).

Ok sounds good.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
-    _metadata_types = {}
-    # ----------------------------------------------------
+    @classmethod
+    def from_image(cls, im):
```
Why not make this a subclass, say `ArrayImageSource`?

vineetbansal replied:
A new class `ArrayImageSource` added.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
 
         :param L: resolution of (square) images (int)
         :param n: The total number of images available
             Note that images() may return a different number of images based on it's arguments.
```
“it's” → “its”.

vineetbansal replied:
changed

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,57 +1,98 @@
-    def __init__(self, L, n, dtype='double', metadata=None):
+    def __init__(self, L, n, dtype='double', metadata=None, memory=None):
         """
         A Cryo-EM Source object that supplies images along with other parameters for image manipulation.
```
“Source” → “`ImageSource`”.

vineetbansal replied:
changed

janden commented on src/aspire/source/__init__.py:
```diff
@@ -64,22 +105,41 @@ def __init__(self, L, n, dtype='double', metadata=None):
+            self._rotations = R.from_euler('ZYZ', self.get_metadata(['_rlnAngleRot', '_rlnAngleTilt', '_rlnAnglePsi']), degrees=True)
+
+        self.model_pipeline = Pipeline(xforms=None, memory=memory)
+        self.generation_pipeline = Pipeline(xforms=None, memory=memory)
```
Why are these not initialized with the values from the metadata fields?

vineetbansal replied:
`model_pipeline` obliterated. Vaporized.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -236,52 +324,56 @@ def cache(self, im=None):
         self._im = im
 
-    def images(self, start=0, num=np.inf, indices=None, *args, **kwargs):
+    def images(self, start=0, num=np.inf, indices=None, batch_size=256, *args, **kwargs):
```
Ideally, batching should be done at the next level up. This is the reason for providing the `(start, num)` indices to `images()`.

vineetbansal replied:
So the reason I added this as an argument:

In the new world where images are run through a `Pipeline` (where one has downsampled the source down the line and is thus confident that (in the extreme case) one can call the `images()` method to get *all* images, without running out of memory), there is still a distinction between how many images the caller wants (in this case *all* of them), vs. how the pipeline processing should be batched. The `batch_size` specifies how to reasonably run the pipeline.

In other words, when saying `src.downsample(8)` followed by `im = src.images()` (which should both work properly when used in conjunction), a `batch_size` is still implied. We can call this something else to make it clear. This will be a non-issue if we eventually end up taking out the whole `Pipeline` out of `images()`.

janden replied:
Ok I see. What I'm worried about here is that we might end up getting two levels of batching (one by the caller of `images()` and one inside of `images()` itself). This would be less efficient than a single level.

The use case of calling `images()` to get *all* the images is quite restrictive, and should ideally only be done during prototyping and testing. Since any general-purpose code *cannot* rely on all the images fitting into memory, it would be written to batch its input as needed. This issue would mostly come up in interactive work or during basic tests (where it should be easy to identify and fix).

A compromise may be to increase the default batch size here (inside `images()`) to something relatively large (say `batch_size=32678`)? That way, it is less likely to affect smaller calls (which may take `n=8192` or comparable).

vineetbansal replied:
Default `batch_size`  increased to `32768`, with the justification in docstrings.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -236,52 +324,56 @@ def cache(self, im=None):
         return im
 
-    def set_max_resolution(self, max_L):
+    def downsample(self, max_L):
```
`max_L` → `L`.

vineetbansal replied:
changed

janden commented on src/aspire/source/__init__.py:
```diff
@@ -236,52 +324,56 @@ def cache(self, im=None):
         Modify the Source object in place by whitening + caching all images, and adding the appropriate whitening
             filter to all available filters.
-        :param whiten_filter: Whitening filter to apply. If None, determined automatically.
+        :param whiten_filter: Whitening filter to apply.
```
Why not allow `ImageSource` to estimate its own whitening filter?

vineetbansal replied:
I thought it was a little random for the whitening operation to assume white noise, but maybe not. Added back in.

janden replied:
Well it's not really appropriate for it to assume white noise (where the noise PSD is constant, and it would not need any whitening), but something like a radially symmetric noise PSD might be reasonable. If that's not available yet, let's leave it without a default for now.

vineetbansal replied:
Default value of `whiten_filter` removed.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -317,27 +402,18 @@ def vol_forward(self, vol, start, num):
-
     def save(self, starfile_filepath, batch_size=1024, overwrite=False):
 
-        df = self._to_starfile_loopdata()
```
I think we discussed this before, but ideally, this `save()` functionality should lie outside the `ImageSource`, since it enforces a specific format (the RELION STAR file format), it's not a generic function of the `ImageSource`.

vineetbansal replied:
A top level `save_star` function added to `aspire.io.star`

janden commented on src/aspire/source/simulation.py:
```diff
@@ -43,6 +49,20 @@ def __init__(self, L=8, n=1024, states=None, filters=None, offsets=None, amplitu
+        # except that we might add a NoiseAdder transform at the end
+        self.generation_pipeline = Pipeline(self.model_pipeline.xforms, memory=self.model_pipeline.memory)
+
+        self.noisy = noisy
```
Isn't it enough just to set the noisy variance to zero to disable the noise?

vineetbansal replied:
There is currently no `noise_variance` parameter when creating a `Simulation`, just a `noise_filter`. But this is a good idea - I've replace `noisy` (boolean) with a `noise_variance` parameter (float), with a default value of 1.0. A value of 0.0 will be equivalent to no noise.

janden replied:
Well it's good to have `noise_filter` specify the noise in the `Simulation`. Having `noise_variance` as well is a bit redundant. However, a `noise_filter` equal to something like `ScalarFilter(0)` (which is maybe its own class, `ZeroFilter()`, in analogy to `IdentityFilter()`), would be equivalent.

vineetbansal replied:
Remove the `noise_variance` parameter from Simulation (just rely on `noise_filter`)

vineetbansal replied:
Added a `ZeroFilter` class. Removed a `noise_variance` parameter from `Simulation` constructor.

janden reviewed (commented):
Some things to discuss here. We talked about a few of them already today, but it's good to get into the specifics.

vineetbansal committed:
Changes pertaining to comments on PR #68

janden reviewed (commented):
For some reason GH is forcing me to make these comments part of a review (part 1).

vineetbansal committed:
More issues related to PR #68 resolved.

Bug fix with `whiten` - A `Simulation` has to do a bit of extra work in addition to what a regular `ImageSource` would do. We were hitherto forcing this extra step (modifying the generation pipeline to the base class).

vineetbansal committed:
Removing some accidentally committed code

vineetbansal committed:
Modified Xform hierarchy to introduce LinearXform.

LinearIndexedXform/LinearPipeline now (multiple)inherit from both LinearXform and their respective parent classes (IndexedXform, Pipeline respectively).
Angstrom -> angstrom as per a PR comment.

vineetbansal committed:
avoiding the use of '3D Image' in comments.

vineetbansal committed:
Changes requested in PR #68.

The `Filter` class now no longer maintains a `power` attribute. A separate `PowerFilter` serves as a thin wrapper around other `Filter` objects.
No longer supporting the concept of a `model_pipeline`, but key operations (amplitude/filtering/shifting) pertaining to running im_backward/vol_forward are performed explicitly when needed.
No longer supporting a `resolution` parameter for `Xform` objects, since the absence of a `model_pipeline` obviates the need to having to downsample individual `Xform` objects.
The `generation_pipeline` is now a `Pipeline` object maintained exclusively by the top level `ImageSource` class. The `Simulation` class or `RelionSource` class do not see/deal with any pipelines, but serve up images appropriately in their `_images` methods.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -316,16 +300,32 @@ def cache(self, im=None):
         self._im = im
 
-    def images(self, start=0, num=np.inf, indices=None, batch_size=256, *args, **kwargs):
+    def images(self, start=0, num=np.inf, indices=None, batch_size=32678, *args, **kwargs):
```
Didn't we ultimately decide to remove `batch_size` and also remove the default values for `start` and `num` (forcing the user to commit to the number of images they want)?

vineetbansal replied:
Ok - maybe I remembered wrong. Removed now.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -341,29 +341,25 @@ def downsample(self, L):
-        :param whiten_filter: Whitening filter to apply.
-        :return: On return, the Source object has been modified in place.
+        Modify the `ImageSource` in-place by appending a whitening filter to the generation pipeline.
+        :param whiten_filter: Whitening filter to apply, as a `Filter` object.
```
So if the user is specifying the *whitening* filter, we do not need to take the negative square root before applying it. If, on the other hand, the user specifies the *noise PSD filter*, we do. So either the input is `noise_filter`, or we keep it as `whiten_filter` (but remove the `PowerFilter`  below).

vineetbansal replied:
Great point. My bad - renamed argument to `noise_filter`

janden commented on src/aspire/source/relion.py:
```diff
@@ -171,22 +169,13 @@ def load_single_mrcs(filepath, df):
 
         logger.info('Adding Whitening Filter Xform to end of generation pipeline')
-        self.generation_pipeline.add_xform(FilterXform(whiten_filter))
+        self.generation_pipeline.add_xform(FilterXform(PowerFilter(whiten_filter, power=-0.5)))
```
Why isn't this handled by `ImageSource.whiten`?

janden replied:
Fixed in next PR.

janden reviewed (commented):
Ok looks good. Just a few comments.

vineetbansal committed:
Removed default values for arguments in the ImageSource.images method.

This is to force the user to think about whether chunking images would be necessary. Not batching calls through the pipeline anymore, but placing onus on the user to know what to do. `whiten_filter` argument renamed as `noise_filter` (internally evaluated at power -0.5)
